### PR TITLE
Enhance that there are two or more values of the same key in a dictio…

### PR DESCRIPTION
**Pull request by shlao**
Updated
```
insights/parsers/__init__.py
```

Here are the codes of parser SystemctlShowServiceAll:
```
 19 class SystemctlShowServiceAll(CommandParser, dict):
 79     def parse_content(self, content): 
 90         for s, e in idx_list:
 91             data = split_kv_pairs(content[s:e], use_partition=False)
 92             name = data.get('Names', data.get('Id'))
 93             if not name:
 94                 raise ParseException('"Names" or "Id" not found!')
 95             self[name] = dict((k, v) for k, v in data.items() if v)
```

Here are the testing data (on RHEL8):
```
# systemctl show sshd.service  > /tmp/sshd.txt
# cat /tmp/sshd.txt | grep Env
EnvironmentFiles=/etc/crypto-policies/back-ends/opensshserver.config (ignore_errors=yes)
EnvironmentFiles=/etc/sysconfig/sshd (ignore_errors=yes)
```
the function split_kv_pairs will overwrite the dictionary value of the same key **_EnvironmentFiles_** and only save the last value.

Signed-off-by: shlao <shlao@redhat.com>

### CODE REVIEW

**Comment by shlao** [comment]
If the key has existed then it is well to combine two dictionary values.
for example, what we get from expression service_conf['sshd.service']['EnvironmentFiles'] is:
`['/etc/crypto-policies/back-ends/opensshserver.config', '/etc/sysconfig/sshd']`

**Comment by shlao** [comment]
@xiangce @bfahr Will you help me to review this?

File: insights/parsers/__init__.py [hunk 1]
```diff
@@ -99,6 +99,75 @@ def make_kv(opt):
     return dict(make_kv(opt) for opt in optlist.split(opt_sep))
 
 
+def set_dict_value(this_dict, key, value):
```
Comment (xiangce) [inline insights/parsers/__init__.py:102]: I think this function makes things more complex.

File: insights/parsers/__init__.py [hunk 2]
```diff
@@ -169,10 +238,10 @@ def split_kv_pairs(lines, comment_char="#", filter_string=None, split_on="=", us
         if not use_partition:
             if split_on in line:
                 k, v = line.split(split_on, 1)
-                kv_pairs[k.strip()] = v.strip()
+                set_dict_value(kv_pairs, k.strip(), v.strip())
         else:
             k, _, v = line.partition(split_on)
-            kv_pairs[k.strip()] = v.strip()
+            set_dict_value(kv_pairs, k.strip(), v.strip())
```
Comment (xiangce) [inline insights/parsers/__init__.py:244]: How about changing it like:
```diff
+        k, v = [None, None]
         if not use_partition:
             if split_on in line:
-                k, v = line.split(split_on, 1)
-                set_dict_value(kv_pairs, k.strip(), v.strip())
+                k, v = [i.strip() for i in line.split(split_on, 1)]
         else:
-            k, _, v = line.partition(split_on)
-            set_dict_value(kv_pairs, k.strip(), v.strip())
+            k, _, v = [i.strip() for i in line.partition(split_on)]
+        if k is not None:
+            if k not in kv_pairs:
+                kv_pairs[k] = v
+            else:
+                _v = kv_pairs[k]
+                _v = [_v] if not isinstance(_v, list) else _v
+                _v.append(v)
```

Comment (shlao) [reply insights/parsers/__init__.py:244]: Let me try it in my host.

**Review by xiangce — COMMENTED** [review]
@shlao  Would you please update the `Examples` for this change as well?

**Review by xiangce — APPROVED** [review]

**Comment by xiangce** [comment]
@bfahr - This change looks good to me, please help double confirm.

**Comment by shlao** [comment]
@bfahr Will you help me to review this PR, thanks!

**Comment by csams** [comment]
The proposed impl would require users to test every value to see whether it's a list or not. I prefer to keep the existing impl and handle the duplicate key use case separately, either in a different function or with a keyword argument to the current one. e.g. if `allow_duplicates` is true, _all_ values should be lists for consistency. Otherwise, go with the current behavior where the last duplicate key wins.

**Comment by xiangce** [comment]
> The proposed impl would require users to test every value to see whether it's a list or not. I prefer to keep the existing impl and handle the duplicate key use case separately, either in a different function or with a keyword argument to the current one. e.g. if `allow_duplicates` is true, _all_ values should be lists for consistency. Otherwise, go with the current behavior where the last duplicate key wins.

I like the `allow_duplicates` idea. Thank you Chris. @shlao how do you think? Would you please update per Chris's proposal?

**Comment by shlao** [comment]
@csams  Good idea. @xiangce  Will add a parameter `allow_duplicates` to make the parser backward-compatible.

**Comment by shlao** [comment]
@csams if all values should be lists for consistency:
```
        >>> split_kv_pairs(lines, allow_duplicates=True)
        {'keyword2': ['value2a=True, value2b=100M'], 'keyword1': ['value1', 'value2']}
```

The the we need to change a lot in parser SystemctlShowServiceAll:
```
 91             data = split_kv_pairs(content[s:e], use_partition=False, allow_duplicates=True)
 92             name = data.get('Names', data.get('Id'))
 95             self[name] = dict((k, v) for k, v in data.items() if v)
```
Line 95 will hit this error
TypeError: unhashable type: 'list'

**Comment by shlao** [comment]
@csams So I change it to store values in a list when there are two or more lines that have the same key
```
                     _v = kv_pairs[k]
                    _v = [_v] if not isinstance(_v, list) else _v
```

**Comment by xiangce** [comment]
> @csams if all values should be lists for consistency:
> 
> ```
>         >>> split_kv_pairs(lines, allow_duplicates=True)
>         {'keyword2': ['value2a=True, value2b=100M'], 'keyword1': ['value1', 'value2']}
> ```
> 
> The the we need to change a lot in parser SystemctlShowServiceAll:
> 
> ```
>  91             data = split_kv_pairs(content[s:e], use_partition=False, allow_duplicates=True)
>  92             name = data.get('Names', data.get('Id'))
>  95             self[name] = dict((k, v) for k, v in data.items() if v)
> ```
> 
> Line 95 will hit this error
> TypeError: unhashable type: 'list'

@shlao - This is because the `name` is also wrapped into a list, it can be resolved by:
```python
            # There must be one and only one "name" as the key
            if not name or len(name) > 1:
                raise ParseException('"Names" or "Id" not found!')
            name = name[0]
            self[name] = dict((k, v) for k, v in data.items() if v)
```

As Chris mentioned, if only store the keys with multiple values into a list, it would require users to test every value to see whether it's a list or not. I don't think this is what we want.

**Comment by shlao** [comment]
@xiangce  @csams  You can see what are changes in latest pr.
Beside that, we need to change the rules using the Parsers which invoke function split_kv_pairs:
```
postgresql_on_external_disk
httpd_fails_to_start_during_boot
sssd_network_user_auth_issue
```

**Comment by xiangce** [comment]
@shlao - Yeap, that's the impact of this update (applying the "allow_duplicates=Ture" to this parser). It's reasonable for me to make this parser more flexible to support these `duplicated keys` (like a bugfix). And we'd better use this parser in this way (I think we should update the existing rules). @bfahr - How do you think?

File: insights/parsers/systemctl_show.py [hunk 3]
```diff
@@ -29,6 +33,8 @@ class SystemctlShowServiceAll(CommandParser, dict):
         LimitNOFILE=65536
         LimitMEMLOCK=
         LimitLOCKS=18446744073709551615
+        DummyKey1 = Value1
+        DummyKey1 = Value2
```
Comment (bfahr) [inline insights/parsers/systemctl_show.py:37]: Is this valid for there to be two keys with the same name for a particular services in the output of this command?

File: insights/parsers/systemctl_show.py [hunk 4]
```diff
@@ -45,6 +51,7 @@ class SystemctlShowServiceAll(CommandParser, dict):
                 "LimitNOFILE"      : "65536",
                 "TimeoutStartUSec" : "1min 30s",
                 "LimitLOCKS"       : "18446744073709551615",
+                "DummyKey1"        : ["Value1", "Value2"]
```
Comment (bfahr) [inline insights/parsers/systemctl_show.py:54]: Based on my comment above, these examples should all be lists.

File: insights/parsers/systemctl_show.py [hunk 5]
```diff
@@ -164,7 +172,7 @@ class SystemctlShowTarget(SystemctlShowServiceAll):
         >>> 'network.target' in systemctl_show_target
         True
         >>> systemctl_show_target.get('network.target').get('WantedBy', None)
-        'NetworkManager.service'
+        ['NetworkManager.service']
```
Comment (bfahr) [inline insights/parsers/systemctl_show.py:175]: Same comment as above, I believe all values will be a `list` type.  Will this command actually return duplicates that require a list?

**Review by bfahr — CHANGES_REQUESTED** [review]

**Comment by shlao** [comment]
I close this one because there is no requirements now.

**Pull request closed without merging**